### PR TITLE
Fix project bv curl conf

### PR DESCRIPTION
**Pull request by rosskynch**
Hi,

Here's a new function to handle the Dirichlet boundary conditions for FE_Nedelec elements. It uses a fairly simple L2-projection method to compute the degrees of freedom on edges and faces. Rather than replace the current function, I've given it a new name as this may be less efficient on geometries with rectangular faces.

This also doesn't solve the issue when the face orientations are not of standard type.

### CODE REVIEW

**Comment by rosskynch** [comment]
Hi, have now merged the hp and non-hp versions and tidied things up a bit. I've also added documentation which hopefully explains how the constraints are computed and why this function might be preferred over the other. It's possible the methods used are actually the same.. I've provided a reference to the paper I used which hopefully is clearer than the book referred to in the original (I have a copy and I've not found the specific section which details the steps).

Anyway, let me know if anything needs improving - I'll squash the commits into one once you're happy!

**Comment by rosskynch** [comment]
@bangerth - just wanted to make sure you'd seen this update - no worries if you've just not been able to get to it.

**Comment by tamiko** [comment]
@bangerth, @rosskynch: What is the status of this pull-request?

**Comment by rosskynch** [comment]
Hi Matthais, I think it's now complete if you're happy with it?

Since there's a merge conflict would it be helpful if I re-forked at the latest version and added the new routines to that?

File: include/deal.II/numerics/vector_tools.h [hunk 1]
```diff
@@ -1381,6 +1381,105 @@ namespace VectorTools
...
+   * i.e. $x$-, $y$-, and finally $z$-component.
+   * 
+   * The parameter @p boundary_component corresponds to the number @p boundary_indicator
+   * of the face. numbers::internal_face_boundary_id is an illegal value, since it is 
+   * reserved for interior faces.
+   * 
+   * The last argument is denoted to compute the normal vector $\vec{n}$ at the
+   * boundary points.
+   * 
+   * <h4>Computing constraints</h4>
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.h]: Move this whole section up, under the place where you have the formula with the E=F. Then separate it from the discussion of the arguments to this function using
```
  <h4>Arguments to this function</h4>
```
or similar.

File: include/deal.II/numerics/vector_tools.h [hunk 2]
```diff
@@ -1381,6 +1381,105 @@ namespace VectorTools
...
+   * 
+   * To compute the constraints we use a projection method based upon the paper mentioned
+   * above. In 2D this is done in a single stage for the edge-based shape functions, regardless
+   * of the order of the finite element. In 3D this is done in two stages, edges first and then
+   * faces.
+   * 
+   * For each cell, each edge, $e$, is projected by solving the linear system Ax=b where
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.h]: dollar signs around Ax=b. But you don't comment on what the x actually correspond to. I suppose the values of the degrees of freedom defined on this cell? I.e., the constraints added to ConstraintMatrix object are x=A^{-1}b?

File: include/deal.II/numerics/vector_tools.templates.h [hunk 3]
```diff
@@ -3818,6 +3818,1561 @@ namespace VectorTools
...
+       * We call the map associated_edge_dof_to_face_dof
+       */
+      std::vector<unsigned int> associated_edge_dof_to_face_dof (degree + 1);
+
+      // Lowest DoF in the base element allowed for this edge:
+      unsigned int lower_bound =
+        fe.base_element(base_indices.first).face_to_cell_index(line * (degree + 1), face);
+      // Highest DoF in the base element allowed for this edge:
+      unsigned int upper_bound =
+        fe.base_element(base_indices.first).face_to_cell_index((line + 1) * (degree + 1) - 1, face);
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.templates.h]: still an issue.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 4]
```diff
@@ -3818,6 +3818,906 @@ namespace VectorTools
...
+              ExcMessage ("Error: Unexpected number of 3D edge DoFs"));
+
+      // Matrix and RHS vectors to store linear system:
+      // We have (degree+1) basis functions for an edge
+      FullMatrix<double> edge_matrix (degree + 1,degree + 1);
+      FullMatrix<double> edge_matrix_inv (degree + 1,degree + 1);
+      Vector<double> edge_rhs (degree + 1);
+      Vector<double> edge_solution (degree + 1);
+
+      Point<dim> tangential;
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.templates.h]: You use the variable only inside the loop, so declare it there (it's cheap to initialize, so no need to pull it out of the loop). As a general guideline, always push the declaration of a variable into the innermost scope and as far down as you can (it makes the code easier to read the more localized the scope of a variable is), unless you have good reasons to pull it out of scopes (e.g., if it requires dynamic memory allocation).

**Comment by bangerth** [comment]
I think part of what makes it hard for me to read the patch is that there are multiple functions in the anonymous namespace with the same name. Can you summarize what they do and how they differ? Does their code have significant overlap?

File: tests/deal.II/nedelec_non_rect_face.cc [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+int main (int argc, char *argv[])
+{
+  using namespace Maxwell;
+
+  for (unsigned int p=0; p<4; ++p)
+    {
+      std::cout << "p = " << p << std::endl;
+      MaxwellProblem<3> (p).run();
+    }
+
```
Comment (bangerth) [inline tests/deal.II/nedelec_non_rect_face.cc]: You'll need to adjust the main function to the usual style we use for testcases (just pick any of the other testcases as an example). Everything you write to then also needs to go to `deallog` instead of `std::cout`.

**Comment by bangerth** [comment]
I still see a few issues, but I think we're getting there :-)

File: tests/deal.II/nedelec_non_rect_face.cc (hunk 5 above)
Comment (guidokanschat) [inline tests/deal.II/nedelec_non_rect_face.cc]: If you use the same style for main as other tests, you should also get output with less decimals. You should use ctest to generate the outputfile and then copy it to the location for comparison

Comment (rosskynch) [inline tests/deal.II/nedelec_non_rect_face.cc]: Can I just clarify how I should set the test up?

Should I write a fe_nedelec_non_rect_face.h with the majority of the code inside and then a short fe_nedelec_non_rect_face.cc ? Or am I ok to keep the test all in one .cc file and just update the main function to attach a logfile and output to deallog, etc?

**Comment by rosskynch** [comment]
Hi again, apologies for the time taken to get back to this. I've tried to include everything you mentioned.

I think I understand your confusion with the multiple internal functions. I've tried to expand a little on why they are there in the comments within them. However, I thought I'd expand on my reasoning for laying it out this way (partly as it was already handled this way by the function I'm supplementing), as well as explaining the method...

Essentially I am handling two problems in one go:

In 2D the constraints need only be applied to basis functions associated with edges. Note the edge *is* a face in 2D, which is why I handle the "edges" inside the internals::compute_face_projection_curl_conforming_l2 function. The basis functions associated with the interior of the cell have no support on the edges of the cell, so there is no need to constrain these DoFs - hence we're finished at this point in 2D.

In 3D you have a similar property with the basis functions associated with the interior of the cell (these are all higher order). However, you do have non-zero values from basis functions associated with both an edge and face, so you must constrain DoFs associated with both. This can be done in two steps, due to the property of the edge and face basis functions. An edge function is non-zero on both the edge and the rest of the face, however a face basis function (it's a "bubble function") is zero on the edge.

At lowest order, we only have edge basis functions, so we only need to deal with them (this is where compute_edge_projection_l2 comes in. At higher orders, you now have basis functions associated with faces as well, so we have to deal with the fact that the constraint on the edge function will make a contribution on the face (this is where the old function is incorrect - it doesn't appear to handle the residuals correctly.. and it's possible things cancel nicely on a rectangular face, which is why it still worked there). So the 3D part of internals::compute_face_projection_curl_conforming_l2 function computes the face projects and also the residual due to the edge functions and computes the adjusted constraints for the boundary condition.

I hope that's clear! Shall update the pull request shortly with my changes - just recompiling and checking I haven't made any mistakes!

File: tests/deal.II/nedelec_non_rect_face.cc (hunk 5 above)
Comment (bangerth) [inline tests/deal.II/nedelec_non_rect_face.cc]: > Should I write a fe_nedelec_non_rect_face.h with the majority of the code
> inside and then a short fe_nedelec_non_rect_face.cc ? Or am I ok to keep the
> test all in one .cc file and just update the main function to attach a logfile
> and output to deallog, etc?

We typically try to keep everything for a test in a single .cc file unless 
there is significant code duplication with other tests.

File: include/deal.II/numerics/vector_tools.h [hunk 6]
```diff
@@ -1381,6 +1381,112 @@ namespace VectorTools
...
+   *
+   * @see @ref GlossBoundaryIndicator "Glossary entry on boundary indicators"
+   */
+  template <int dim, class DH>
+  void project_boundary_values_curl_conforming_l2 (const DH &dof_handler,
+                                                   const unsigned int first_vector_component,
+                                                   const Function<dim,double> &boundary_function,
+                                                   const types::boundary_id boundary_component,
+                                                   ConstraintMatrix &constraints,
+                                                   const hp::MappingCollection<dim, dim> &mapping_collection = hp::StaticMappingQ1<dim>::mapping_collection);
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.h]: There are only two kinds of DoFHandlers in deal.II -- hp and non-hp. Since you handle the non-hp case below, just remove the second template argument from this function and make it take a hp::DoFHandler as first argument. Then let both this and the other function call an internal function in the .cc file that does what this function currently does.

It's not a big deal but I've found that if something raises a question ("what kinds of DoFHandler can I pass to this function") then one of my students will have this question and it will take 20 minutes to explain :-) The point obviously being that if one of my students has a question, then 10 more around the world will have the same question.

**Comment by bangerth** [comment]
Thanks, Ross, for the explanation and the tenacity with this! Thanks in particular for the excellent testcase that tests the entire Nedelec-related collection of functions!

Can you address the one issue with the template argument? You'll also need to rebase to master to allow for the patch to be merged. Please let me know when you're done and I'll push the button.

**Comment by rosskynch** [comment]
Hi, made the final change and have done a rebase. I've not done a rebase master with that huge a time difference before so I hope I didn't break anything!

**Comment by bangerth** [comment]
OK, truly fantastic. I really appreciate your willingness to work with us through the various iterations -- many thanks for the work and the tenacity!

**Comment by agrayver** [comment]
@rosskynch, great job, thanks! I wonder if what you have done here can be used to fix FE_Nedelec::interpolate? see: https://github.com/dealii/dealii/issues/328

**Comment by rosskynch** [comment]
@agrayver Thanks for bringing my attention to it, I'll try to take a look soon!

**Comment by bangerth** [comment]
@rosskynch : Since you know this issue best, would you mind taking a look at issue #642?
